src/train.py

```diff
@@ -61,6 +96,179 @@ def train(args):
+            losses.append(loss.detach().cpu().item())
+            if train_log is not None:
+                train_log.add_scalar("loss", losses[-1], global_step)
+            global_step += 1
+
+        # evaluate(model, valid_dataset)
+
+        avg_l = np.mean(losses)
+        print("epoch %-3d \t loss = %0.3f \t" % (epoch, avg_l))
+        # if validation < best:
+        # TODO: Add model (and optimizer) saving for reloading training
+        #     print("Saving model!")
+        #     best = validation
+        #     model.save_model()
+    return model
+
+def train_speech_auto(args):
+    '''
+    TODO:
+        1. Get Dataset
+        2. Init models & optimizers
+        3. Train, or for each epoch:
+                For each batch:
+                - Choose which loss function
+                - Freeze appropriate networks
+                - Run through networks
+                - Get loss
+                - Update
+                Metrics (like validation, etc)
+        4. Return trained text & speech for validation, metric measurement
+
+        TODO: Include functionality for saving, loading from save
+    '''
+    set_seed(args.seed)
+
+    # TODO: Replace get_dataset() with getting train/valid/test split
+    dataset = get_dataset()
+    #NOTE: Subset for prototyping
+    # dataset = torch.utils.data.Subset(dataset, range(1000))
+    dataloader = DataLoader(dataset, batch_size=args.batch_size, shuffle=True, collate_fn=collate_fn_transformer, drop_last=True, num_workers=16)
+    train_log, valid_log = None, None
+    model = SpeechRNN(args).to(DEVICE)
+    print("Sent model to", DEVICE)
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr,  weight_decay=1e-5)
+
+    global_step = 0
+
+    for epoch in range(args.epochs):
+        losses = []
+        for data in dataloader:
+            character, mel, mel_input, pos_text, pos_mel, text_len = data
+            mel_input = mel_input.to(DEVICE)
+            encoder_outputs, latent_hidden, pad_mask = model.encode(mel)
+            pred, stop_pred = model.decode_sequence(mel_input, latent_hidden, encoder_outputs, pad_mask)
+            
+            loss = F.mse_loss(pred, mel)
+            
+            optimizer.zero_grad()
+            loss.backward()
+            if args.grad_clip > 0.0:
+                nn.utils.clip_grad_norm_(model.parameters(), args.grad_clip)
+            optimizer.step()
+            losses.append(loss.detach().cpu().item())
+            if train_log is not None:
+                train_log.add_scalar("loss", losses[-1], global_step)
+            global_step += 1
+
+        # evaluate(model, valid_dataset)
+
+        avg_l = np.mean(losses)
+        print("epoch %-3d \t loss = %0.3f \t" % (epoch, avg_l))
+        # if validation < best:
+        # TODO: Add model (and optimizer) saving for reloading training
+        #     print("Saving model!")
+        #     best = validation
+        #     model.save_model()
+    return model
+
+def train(args):
```

Comment:
Probably dead code here, this function and its contents can be removed. The real `train()` loop is below this one. From what I'm seeing, you're going to call `train_text_auto()` and `train_speech_auto()` in the train loop, right?

Probably change `train_*_auto()` to include a dataloader parameter to pass the unsupervised dataloader to them.

Are we going to have separate functions for cross-modal training?

Reply:
The train_text_auto() and train_speech_auto are purely for ensuring the text and speech autoencoder work independently rn. They will be removed once train() is done, and replaced with something like `text_auto_step` which takes a batch and the text model and does one iteration of text_autoencoder training